### PR TITLE
[Reviewer: Ellie] Big PR of split-storage changes

### PR DESCRIPTION
**Pull request by johadalin**
It may be best for you to have a glance over it all, and then I can tell you why i did each bit as i did, or if it would fit you better I could add line notes where there are notable changes? 

Let me know how you want to go about handling this.

### CODE REVIEW

File: cookbooks/clearwater/recipes/database.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# @file database.rb
+#
+# Project Clearwater - IMS in the Cloud
+# Copyright (C) 2013  Metaswitch Networks Ltd
```
Comment (eleanor-merry) [inline cookbooks/clearwater/recipes/database.rb]: !!!

File: cookbooks/clearwater/recipes/shared_config.rb [hunk 2]
```diff
@@ -63,6 +63,17 @@
   ""
 end
 
+if node[:clearwater][:split_storage]
+  database = "database#{site_suffix}.#{domain}"
```
Comment (eleanor-merry) [inline cookbooks/clearwater/recipes/shared_config.rb]: I don't think this should be site_suffix (as there's no GR support yet)

File: cookbooks/clearwater/templates/default/database_local_config.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+<% if not @node[:clearwater][:index].nil? %>node_idx=<%= @node[:clearwater][:index] %><% end %>
+
+# Local IP configuration
+local_ip=<%= @node[:cloud][:local_ipv4] %>
+public_ip=<%= @node[:cloud][:public_ipv4] %>
+public_hostname=<%= @node[:cloud][:public_hostname] %>
+
+# GR site names
+local_site_name=<%= @local_site %>
+remote_site_name=<%= @remote_site %>
```
Comment (eleanor-merry) [inline cookbooks/clearwater/templates/default/database_local_config.erb]: I'd remove the remote_site_name

File: cookbooks/clearwater/templates/default/shared_config.erb [hunk 4]
```diff
@@ -10,6 +10,22 @@ sas_server=<%= @node[:clearwater][:sas_server] or "0.0.0.0" %>
 enum_server=<%= @node[:clearwater][:enum_server] or "" %>
 alias_list=<%= @alias_list %>
 
+<% if not @database.nil? %>
```
Comment (eleanor-merry) [inline cookbooks/clearwater/templates/default/shared_config.erb]: Why not just 'if @database' (or at least if @database.nil) - it's easier to read

File: cookbooks/clearwater/templates/default/shared_config.erb [hunk 5]
```diff
@@ -10,6 +10,22 @@ sas_server=<%= @node[:clearwater][:sas_server] or "0.0.0.0" %>
...
+# Storage Pool
+cassandra_hostname=<%= @database %>
+chronos_hostname=<%= @database %>
+ralf_session_store=single_site=<%= @database %>
```
Comment (eleanor-merry) [inline cookbooks/clearwater/templates/default/shared_config.erb]: Do you not need quote marks around these values?

File: cookbooks/clearwater/templates/default/shared_config.erb [hunk 6]
```diff
@@ -10,6 +10,22 @@ sas_server=<%= @node[:clearwater][:sas_server] or "0.0.0.0" %>
 enum_server=<%= @node[:clearwater][:enum_server] or "" %>
 alias_list=<%= @alias_list %>
 
+<% if not @database.nil? %>
+# Storage Pool
+cassandra_hostname=<%= @database %>
+chronos_hostname=<%= @database %>
```
Comment (eleanor-merry) [inline cookbooks/clearwater/templates/default/shared_config.erb]: Do any of these values need ports?

File: plugins/knife/clearwater-dns-records.rb [hunk 7]
```diff
@@ -195,27 +195,40 @@ def dns_records
...
+  homestead_dns = {
+    "hs" => {
       :type  => "A",
-      :value => ipv4s_local_site1(find_active_nodes("homestead")),
+      :value => ipv4s_local(find_active_nodes("homestead")),
       :ttl   => "60"
     },
 
-    "homer-site1" => {
+   "hs-site1" => {
```
Comment (eleanor-merry) [inline plugins/knife/clearwater-dns-records.rb]: Alignment

File: plugins/knife/clearwater-dns-records.rb [hunk 8]
```diff
@@ -224,18 +237,43 @@ def dns_records
       :value => ipv4s_local_site2(find_active_nodes("homestead")),
       :ttl   => "60"
     },
+  }
 
-    "homer-site2" => {
+  ralfstead_dns = {
+    "hs" => {
```
Comment (eleanor-merry) [inline plugins/knife/clearwater-dns-records.rb]: I don't like this being called hs

Comment (eleanor-merry) [inline plugins/knife/clearwater-dns-records.rb]: Oh wait, I understand what this is now

File: plugins/knife/knife-deployment-resize.rb [hunk 9]
```diff
@@ -118,18 +118,17 @@ class DeploymentResize < Chef::Knife
...
       return result
     end
 
     def update_ralf_hostname environment, cloud
       ralfs = find_nodes(roles: "chef-base", role: "ralf").length
-
       changed_nodes = []
 
-      %w{bono ibcf sprout ralf}.each do |node_type|
+      %w{bono ibcf sprout ralf ralfstead}.each do |node_type|
```
Comment (eleanor-merry) [inline plugins/knife/knife-deployment-resize.rb]: I don't think this needs doing for ralfsteads, no?

File: plugins/knife/knife-deployment-resize.rb [hunk 10]
```diff
@@ -199,9 +198,15 @@ def run
         ibcf: config[:ibcf_count] || old_counts[:ibcf],
         sipp: config[:sipp_count] || old_counts[:sipp],
         seagull: seagull_count || old_counts[:seagull] }
+        if attributes["split_storage"]
+          new_counts[:database] = config[:database_count] || [old_counts[:database], 1].max
+          new_counts.delete(:homestead)
```
Comment (eleanor-merry) [inline plugins/knife/knife-deployment-resize.rb]: Delete ralf as well?

File: plugins/knife/knife-deployment-resize.rb [hunk 11]
```diff
@@ -249,28 +267,34 @@ def run
                             "chef_environment:#{config[:environment]}")
 
         # Create and upload the shared configuration. This should just be done
-        # on a single node in each site. We choose the first Sprout.
-        sprouts = find_nodes(roles: 'sprout')
-        sprouts.sort_by! { |n| n[:clearwater][:index] }
+        # on a single node in each site. We choose the first Sprout or Database,
```
Comment (eleanor-merry) [inline plugins/knife/knife-deployment-resize.rb]: More GR nastiness here

File: roles/ralfstead.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
+# @file ralfstead.rb
+#
+# Project Clearwater - IMS in the Cloud
+# Copyright (C) 2013  Metaswitch Networks Ltd
```
Comment (eleanor-merry) [inline roles/ralfstead.rb]: Please fix these throughout

**Comment by eleanor-merry** [comment]
This generally looks good, I've made some minor comments inline

File: cookbooks/clearwater/templates/default/shared_config.erb (hunk 4 above)
Comment (johadalin) [inline cookbooks/clearwater/templates/default/shared_config.erb]: I was following the format as later in the file, e.g. 
```
# HSS configuration
 <% if not @hss.nil? %>hss_hostname=<%= @hss %>...
```
I wasn't sure if there was some idiosyncratic reason one couldn't just check `if @database`; does just checking `if @database` see if there is a `database` variable, as opposed to whether it has a real value assigned?

File: cookbooks/clearwater/templates/default/shared_config.erb (hunk 6 above)
Comment (johadalin) [inline cookbooks/clearwater/templates/default/shared_config.erb]: @richardwhiuk , can you comment on this? 
I took the config as seen on your live nodes, and don't think these need ports, but i may be wrong

File: cookbooks/clearwater/templates/default/shared_config.erb (hunk 5 above)
Comment (johadalin) [inline cookbooks/clearwater/templates/default/shared_config.erb]: I don't believe so. 
Again @richardwhiuk , can you please comment on this?

File: plugins/knife/knife-deployment-resize.rb (hunk 9 above)
Comment (johadalin) [inline plugins/knife/knife-deployment-resize.rb]: I think this tails over from when i was trying to plumb the ralf attribute through in split-storage land. 
I'm going to make it just skip the setting has_ralf step if split_storage. Sound reasonable?

File: cookbooks/clearwater/templates/default/shared_config.erb (hunk 6 above)
Comment (richardwhiuk) [inline cookbooks/clearwater/templates/default/shared_config.erb]: Nope these are purely hostnames - the port is hardcoded.

File: cookbooks/clearwater/templates/default/shared_config.erb (hunk 5 above)
Comment (richardwhiuk) [inline cookbooks/clearwater/templates/default/shared_config.erb]: No, you don't - you can't include spaces in the databases parameter

**Comment by johadalin** [comment]
@eleanor-merry Could you just take a look at the changes to knife-deployment-resize from the last commit to see that you're happy with how i've handled the abort and the nastiness around shared_config. 

Tested, and it's fine with both types of deployments, and aborts safely and correctly

**Comment by eleanor-merry** [comment]
LGTM

**Pull request merged**